File: scripts/localization/Lang.cs
using System.Collections.Generic;
using System.Text.Json;
using Godot;

namespace ProiectDeLegume.Scripts.Localization;

public static class Lang
{
    private static Dictionary<string, string> _strings = new();

    public static void Load(string locale = "en")
    {
        var file = FileAccess.Open($"res://assets/data/lang_{locale}.json", FileAccess.ModeFlags.Read);
        if (file == null)
        {
            GD.PrintErr($"Cannot open lang_{locale}.json");
            return;
        }

        string jsonText = file.GetAsText();
        file.Close();

        var dict = JsonSerializer.Deserialize<Dictionary<string, string>>(jsonText);
        if (dict != null) _strings = dict;

        GD.Print($"Loaded {_strings.Count} strings for locale '{locale}'");
    }

    public static string Get(string key)
    {
        return _strings.TryGetValue(key, out var value) ? value : key;
    }

    public static string Get(string key, params (string placeholder, string value)[] args)
    {
        if (!_strings.TryGetValue(key, out var result)) return key;
        foreach (var (placeholder, value) in args)
        {
            result = result.Replace($"{{{placeholder}}}", value);
        }
        return result;
    }
}

File: src/commonMain/kotlin/localization/Lang.kt
package localization

import korlibs.io.file.std.*
import kotlinx.serialization.json.*

object Lang {
    private var strings = mapOf<String, String>()

    suspend fun load(locale: String = "en") {
        val jsonText = resourcesVfs["data/lang_$locale.json"].readString()
        val jsonObj = Json.parseToJsonElement(jsonText).jsonObject
        strings = jsonObj.mapValues { it.value.jsonPrimitive.content }
    }

    operator fun get(key: String): String = strings[key] ?: key

    fun get(key: String, vararg args: Pair<String, String>): String {
        var result = strings[key] ?: return key
        for ((placeholder, value) in args) {
            result = result.replace("{$placeholder}", value)
        }
        return result
    }
}
